Permettre de suivre les contenus d'un membre

GerardPaligot opened the pull request:
| Q | R |
| --- | --- |
| Type de modification | nouvelle fonctionnalité / évolution |
| Ticket(s) (_issue(s)_) concerné(s) | ~ZEP-13 |

Attention : Cette PR est un cadeau pour @gustavi pour sa ZEP (https://github.com/zestedesavoir/zds-site/pull/3606), pensez à merger cette PR avant la sienne ! :)
# QA
- Connectez-vous avec un membre X.
- Rédigez un contenu et demandez sa publication.
- Connectez-vous avec un membre du staff Y.
- Rendez-vous sur le profil du membre et suivez ses nouveaux contenus (sidebar).
- Acceptez la publication du membre X.
- Vérifiez que vous avez bien une notification.
- Cliquez dessus, vérifiez que vous êtes redirigé sur le contenu et vérifiez que la notification est marquée comme lue.
## QA supplémentaire

Donnez une attention particulière aux points suivants :
- Vous pouvez suivre les publications depuis un bouton de la sidebar d'un membre sur sa page de profil, depuis sa liste des tutoriels ou depuis sa liste des articles.
- Si vous n'êtes pas connecté, vous ne pouvez pas suivre les publications d'un membre. Les boutons ne sont pas affichés sur aucune des 3 pages.
- Vous pouvez suivre par e-mail les publications d'un membre.
- Un membre suit ses propres publications. Vous ne recevez donc plus un MP à une publication mais une notification.
- Vérifiez que vous pouvez toujours suivre un forum, un sujet et un contenu depuis l'interface.
- Vérifiez que le nombre de souscription est bien affiché pour un sujet, un forum, un contenu et les nouvelles publications.

Enjoy! :)

GerardPaligot committed:
feat(notif): Permet la souscription a un contenu.

Un utilisateur peut souscrire aux contenus de tous les autres
membres a travers le menu sur la page qui liste le contenu
d'un membre. Il peut souscrire par le centre de notification
zds et recevoir un e-mail pour chaque publication de
l'utilisateur si c'est souhaite.

ZEP-13

GerardPaligot committed:
feat(notif): Genere et marque comme lu une notif sur un contenu.

Quand un membre du staff publie un contenu, une notification est
genere pour toutes les personnes qui ont souscris aux auteurs de
ce contenu. La notification redirige vers la version en ligne de
ce meme contenu et est marque comme lue lorsqu'on affiche la
page de cetet version en ligne.

ZEP-13

GerardPaligot committed:
feat(notif): Genere une notif pour l'auteur.

Avant ce commit, lorsqu'un contenu est publie, la
plateforme genere un MP pour signaler aux auteurs que
le contenu a ete publie. Maintenant, une notification
est genere parce que les auteurs souscrivent sans le
savoir a leur propre actualite.

ZEP-13

GerardPaligot committed:
feat(notif): Permet de suivre des publications a pls endroits.

Un utilisateur connecte peut suivre les nouvelles publications
d'un membre a partir de la liste de ses publications publiques
et a partir de son profil.

ZEP-13

GerardPaligot committed:
fix(flake): Imports.

GerardPaligot committed:
refactor(template): Utilise les templates pour suivre un forum.

gustavi commented:
Tout plein de ❤️ !

gustavi commented on templates/tutorialv2/index.html:
```diff
@@ -112,6 +113,37 @@ <h2 class="ico-after ico-tutorials">
 {% endblock %}
 
+{% block sidebar_actions %}
+    {% if usr != request.user and request.user.is_authenticated %}
```
Tu peux changer `request.user` par `user` (le premier utilise le context processor `django.core.context_processors.request` et le second `django.contrib.auth.context_processors.auth`).

GerardPaligot committed:
refactor(template): Utilise les templates pour suivre un sujet.

GerardPaligot committed:
fix(template): Utilise user plutot que request.user.

GerardPaligot committed:
refactor(template): Utilise le template pour suivre un contenu.

GerardPaligot committed:
fix(email): Ajoute les templates e-mail pour new publications.

Quand un membre suit un autre membre pour recevoir des alertes
quand il publie de nouvelles publications et qu'il desire recevoir
un e-mail a chaque fois, il faut creer des templates qui sera
utilise pour le corps de ces mails.

GerardPaligot committed:
feat(souscription): Affiche publiquement le nombre de souscription.

gustavi commented on templates/tutorialv2/index.html:
```diff
@@ -112,6 +113,40 @@ <h2 class="ico-after ico-tutorials">
+    <li class="inactive">
+        {% include 'notification/subscription_count_template.html' with subscriber_count=subscriber_count %}
+    </li>
+    {% if usr != user and request.user.is_authenticated %}
```
request.user => user

vhf commented on zds/tutorialv2/views/views_validations.py:
```diff
@@ -1,5 +1,7 @@
 # coding: utf-8
+import logging
```
Est-ce nécessaire ?

GerardPaligot replied:
Il a été déplacé automatiquement par PyCharm.

GerardPaligot replied:
see https://github.com/zestedesavoir/zds-site/pull/3694/files#diff-f2064ec573e3eb343c9c83f99dc36109L25

gustavi replied:
Ça fait pas de mal de ranger un peu les imports (j'essaye de le faire aussi et je me rends souvent compte de choses pas terribles comme des doubles imports).

Emeric54 commented:
Une PR sympa, merci :+1: 

**QA =>** :ok: !

Toutefois, j'ai juste une petite chose qui me gêne : 

> Un membre suit ses propres publications. Vous ne recevez donc plus un MP à une publication mais une notification.

Depuis récemment, le MP envoyé lors de la publication permettait de transmettre le commentaire du validateur. Avec le nouveau fonctionnement, on perd ce commentaire, ce qui est dommage.

GerardPaligot commented:
Un grand merci pour ta QA !

> Depuis récemment, le MP envoyé lors de la publication permettait de transmettre le commentaire du validateur. Avec le nouveau fonctionnement, on perd ce commentaire, ce qui est dommage.

Toutafé et j'y ai pensé avant de retirer la fonctionnalité. Pourquoi ? De mon expérience d'auteur sur ZdS, ce commentaire n'a jamais servi. La publication se contente de publier le contenu après une discussion soutenue entre le validateur et le(s) auteur(s).

Après, je peux retirer la souscription et remettre le message privé. C'est un seul commit à retirer. :)

Emeric54 commented:
Une discussion similaire avait déjà eue lieu lors de la mise en place de l'envoi du commentaire du validateur. Je suis toujours pour cet envoi, notamment car : 
- il est entré par le validateur et enregistré, autant s'en servir ;
- un petit merci, ça fait jamais de mal ;
- et pour moi, la publication d'un des ses contenus est plus importante qu'une simple notification.

GerardPaligot commented:
> Une discussion similaire

Si tu as un lien, je suis preneur pour relancer la discussion.

Emeric54 commented:
> Si tu as un lien, je suis preneur pour relancer la discussion.

Je viens de vérifier, en fait il s'agissait juste d'une proposition que j'avais faite sur GH, à laquelle personne ne s'était opposé, donc pas vraiment une discussion...

Emeric54 commented:
Les validos se servent en fait du commentaire de validation plutôt en interne apparement, du coup, pas besoin de l'envoyer, tu peux laisser la notif.

Du coup, je crois que la PR est toute bonne !

GerardPaligot commented:
Parfait. Ready for merge alors! cc @zestedesavoir/les-jeunes-pousses
